sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -0,0 +1,175 @@
+    return MContext.get_platform().get_backend();
+  }
+
+  context get_context() const noexcept { return MContext; }
+
+  std::vector<device> get_devices() const noexcept { return MDevices; }
+
+  std::vector<kernel_id> get_kernel_ids() const {
+    // Collect kernel ids from all device images, then remove duplicates
+
+    std::vector<kernel_id> Result;
+    for (const device_image_plain &DeviceImage : MDeviceImages) {
+      const std::vector<kernel_id> &KernelIDs =
+          getSyclObjImpl(DeviceImage)->get_kernel_ids();
+
+      Result.insert(Result.end(), KernelIDs.begin(), KernelIDs.end());
+    }
+    std::sort(Result.begin(), Result.end(),
+              [](const kernel_id &LHS, const kernel_id &RHS) {
+                return detail::getSyclObjImpl(LHS) <
+                       detail::getSyclObjImpl(RHS);
+              });
+    std::unique(Result.begin(), Result.end());
+
+    return Result;
+  }
+
+  kernel get_kernel(const kernel_id &KernelID) const {
+    throw sycl::runtime_error("Not implemented", PI_INVALID_OPERATION);
+  }
+
+  bool has_kernel(const kernel_id &KernelID) const noexcept {
+    return std::any_of(MDeviceImages.begin(), MDeviceImages.end(),
+                       [&KernelID](const device_image_plain &DeviceImage) {
+                         return DeviceImage.has_kernel(KernelID);
+                       });
+  }
+
+  bool has_kernel(const kernel_id &KernelID, const device &Dev) const noexcept {
+    return std::any_of(
+        MDeviceImages.begin(), MDeviceImages.end(),
+        [&KernelID, &Dev](const device_image_plain &DeviceImage) {
+          return DeviceImage.has_kernel(KernelID, Dev);
+        });
+  }
+
+  bool contains_specialization_constants() const noexcept {
+    return std::any_of(
+        MDeviceImages.begin(), MDeviceImages.end(),
+        [](const device_image_plain &DeviceImage) {
+          return getSyclObjImpl(DeviceImage)->has_specialization_constants();
+        });
+  }
+
+  bool native_specialization_constant() const noexcept {
+    return std::all_of(MDeviceImages.begin(), MDeviceImages.end(),
+                       [](const device_image_plain &DeviceImage) {
+                         return getSyclObjImpl(DeviceImage)
+                             ->all_specialization_constant_native();
+                       });
+  }
+
+  bool has_specialization_constant(unsigned int SpecID) const noexcept {
+    return std::any_of(MDeviceImages.begin(), MDeviceImages.end(),
+                       [SpecID](const device_image_plain &DeviceImage) {
+                         return getSyclObjImpl(DeviceImage)
+                             ->has_specialization_constant(SpecID);
+                       });
+  }
+
+  void set_specialization_constant_raw_value(unsigned int SpecID,
+                                             const void *Value,
+                                             size_t ValueSize) {
+    for (const device_image_plain &DeviceImage : MDeviceImages)
```

Comment:
I wonder if some synchronization is needed here. What if this kernel bundle is JITted in another thread while this copying spec constant value(s) is in progress?

Reply:
Agree, this can be a problem, will add a lock to avoid reads of "partial writes".
@gmlueck  FYI

Reply:
Agreed that this makes sense.

Reply:
I saw that there was synchronization added in `device_image_impl`.

Reply:
@gmlueck - thanks for the pointer. LGTM then.